docs/design.md

```diff
@@ -0,0 +1,153 @@
+# Design
+
+## Models
+
+A Model object represents a logical ML model. The object describes a location
+where ML model files can be stored for a specific model. 
+
+Optionally, a loader can be specified to load ML models weights and biases from specific sources.
+For example, a commonly used model loader would be the HuggingFace model loader
+where by providing the model name would load the model to the Model object
+bucket location. Model loaders are defined as container images and can either
+point to prebuilt container images or a Git repository with Dockerfile.
+
+A model
+
+The Model object can optionally describe the training parameters including a reference to the Dataset that should be used for training. In this case, the Model object represents the eventual output of a training/finetuning operation.
+
+```yaml
+kind: Model
+spec:
+  container: # used for either loader or trainer
+  loader: {} # Optional, when empty an just a location will be created with no files in it
+  trainer: {} # Optional, will run a training job
+```
+
+### Model Sources
+
+A Model can be sourced from Git or a prebuilt image. Using Git will trigger a container build Job.
+
+```yaml
+kind: Model
+name: falcon-7b
+spec:
+  container:
+    git:
+      url: https://github.com/substratusai/hf-model-loader
+    image: substratusai/hf-model-loader
+  loader:
+    params: {name: tiuu/falcon-7b}
+    # optional future
+    paramsFrom: configMap?
+```
+
+`params` get converted to environment variables following a specific scheme
+
+The controller will orchestrate the following flow in this case:
+
+<img src="./diagrams/model-building.excalidraw.png" width="80%"></img>
+
+
+### Model Training
+
+Approach 1: Include in Model object
+
+Models can be trained by specifying the `.spec.training` section.
+
+This will orchestrate a training Job with multiple FUSE mounted buckets.
+
+<img src="./diagrams/model-training.excalidraw.png" width="80%"></img>
+
+```yaml
+kind: Model
+name: falcon-7b-k8s
+spec:
+  container:
+  # can only specify git or image not both
+    git: 
+      url: https://github.com/substratusai/hf-llm-trainer
+      ref: refs/branch/main
+    image: substratusai/hf-llm-peft-trainer # eventually introduce native pytorch trainer
+
+  trainer:
+    # sourceModel can be left out for creating new base model
+    sourceModel:
+      name: falcon-7b
+      namespace: other-namespace
+    dataset:
+      name: k8s-instructions
+      namespace: other-namespace
+    resources: {}
```

Comment:
should resources be under `container`?

Reply:
Currently everything under container is a source of the container... this would break that pattern. I vote `.spec.resources` because that way we can keep it consistent across all object kinds.